enterprise/cli/dbcrypt_rotate.go

```diff
@@ -0,0 +1,152 @@
+	"cdr.dev/slog/sloggers/sloghuman"
+	"github.com/coder/coder/v2/cli"
+	"github.com/coder/coder/v2/cli/clibase"
+	"github.com/coder/coder/v2/coderd/database"
+	"github.com/coder/coder/v2/codersdk"
+	"github.com/coder/coder/v2/enterprise/dbcrypt"
+
+	"golang.org/x/xerrors"
+)
+
+func (*RootCmd) dbcryptRotate() *clibase.Cmd {
+	var (
+		vals = new(codersdk.DeploymentValues)
+		opts = vals.Options()
+	)
+	cmd := &clibase.Cmd{
+		Use:   "dbcrypt-rotate --postgres-url <postgres_url> --external-token-encryption-keys <new-key>,<old-keys>",
+		Short: "Rotate database encryption keys",
+		Options: clibase.OptionSet{
+			*opts.ByName("Postgres Connection URL"),
+			*opts.ByName("External Token Encryption Keys"),
+		},
+		Middleware: clibase.Chain(
+			clibase.RequireNArgs(0),
+		),
+		Handler: func(inv *clibase.Invocation) error {
+			ctx, cancel := context.WithCancel(inv.Context())
+			defer cancel()
+			logger := slog.Make(sloghuman.Sink(inv.Stdout))
+			if ok, _ := inv.ParsedFlags().GetBool("verbose"); ok {
+				logger = logger.Leveled(slog.LevelDebug)
+			}
+
+			if vals.PostgresURL == "" {
+				return xerrors.Errorf("no database configured")
+			}
+
+			switch len(vals.ExternalTokenEncryptionKeys) {
+			case 0:
+				return xerrors.Errorf("no external token encryption keys provided")
+			case 1:
+				logger.Info(ctx, "only one key provided, data will be re-encrypted with the same key")
+			}
+
+			keys := make([][]byte, 0, len(vals.ExternalTokenEncryptionKeys))
+			var newKey []byte
+			for idx, ek := range vals.ExternalTokenEncryptionKeys {
+				dk, err := base64.StdEncoding.DecodeString(ek)
+				if err != nil {
+					return xerrors.Errorf("key must be base64-encoded")
+				}
+				if idx == 0 {
+					newKey = dk
+				} else if bytes.Equal(dk, newKey) {
+					return xerrors.Errorf("old key at index %d is the same as the new key", idx)
+				}
+				keys = append(keys, dk)
+			}
+
+			ciphers, err := dbcrypt.NewCiphers(keys...)
+			if err != nil {
+				return xerrors.Errorf("create ciphers: %w", err)
+			}
+
+			sqlDB, err := cli.ConnectToPostgres(inv.Context(), logger, "postgres", vals.PostgresURL.Value())
+			if err != nil {
+				return xerrors.Errorf("connect to postgres: %w", err)
+			}
+			defer func() {
+				_ = sqlDB.Close()
+			}()
+			logger.Info(ctx, "connected to postgres")
+
+			db := database.New(sqlDB)
+
+			cryptDB, err := dbcrypt.New(ctx, db, ciphers...)
+			if err != nil {
+				return xerrors.Errorf("create cryptdb: %w", err)
+			}
+
+			users, err := cryptDB.GetUsers(ctx, database.GetUsersParams{})
+			if err != nil {
+				return xerrors.Errorf("get users: %w", err)
+			}
+			logger.Info(ctx, "encrypting user tokens", slog.F("count", len(users)))
+			for idx, usr := range users {
+				err := cryptDB.InTx(func(tx database.Store) error {
+					userLinks, err := tx.GetUserLinksByUserID(ctx, usr.ID)
+					if err != nil {
+						return xerrors.Errorf("get user links for user: %w", err)
+					}
+					for _, userLink := range userLinks {
```

Comment:
I think you should log every ID touched at debug level, and every error message should contain the ID of the row that failed